File: 백준/Gold/4485. 녹색 옷 입은 애가 젤다지？/녹색 옷 입은 애가 젤다지？.py
# 녹색 옷 입은 애가 젤다지?
import sys
input = sys.stdin.readline
from collections import deque

direc = [(0, 1), (1, 0), (0, -1), (-1, 0)]
p = 0
while True:
    n = int(input())

    if n == 0:
        break
    p += 1
    graph = []

    for i in range(n):
        graph.append(list(map(int,input().split())))

    visited = [[float('inf')] * n for _ in range(n)]

    que = deque([(0, 0, graph[0][0])])

    while que:
        cr, cc, cnt = que.popleft()

        if visited[cr][cc] < cnt:
            continue
        
        for r, c in direc:
            nr = cr + r
            nc = cc + c

            if not (0 <= nr < n) or not (0 <= nc < n):
                continue

            if graph[nr][nc] + cnt >= visited[nr][nc]:
                continue
        
            visited[nr][nc] = cnt + graph[nr][nc]
            que.append((nr, nc, cnt + graph[nr][nc]))
                
    print(f'Problem {p}: {visited[n-1][n-1]}')
        

File: 백준/Gold/4485. 녹색 옷 입은 애가 젤다지？/녹색 옷 입은 애가 젤다지？.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        int[][] direction = {{0,1},{0,-1},{1,0},{-1,0}};
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int p = 0;
        while (true) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken());

            if (n == 0) {
                break;
            }

            p++;
            int[][] graph = new int[n][n];
            int[][] visited = new int[n][n];

            LinkedList<int[]> que = new LinkedList<>();

            for (int i = 0; i < n; i++) {
                st = new StringTokenizer(br.readLine());
                for (int j = 0; j < n; j++) {
                    graph[i][j] = Integer.parseInt(st.nextToken());
                    visited[i][j] = Integer.MAX_VALUE;
                }
            }

            que.add(new int[] {0, 0, graph[0][0]});

            while (!que.isEmpty()) {
                int[] cur = que.pollFirst();
                int cr = cur[0];
                int cc = cur[1];
                int cost = cur[2];

                if (cost > visited[cr][cc]) {
                    continue;
                }

                for (int[] d : direction) {
                    int nr = cr + d[0];
                    int nc = cc + d[1];

                    if (!(0 <= nr && nr < n) || !(0 <= nc && nc <n )) {
                        continue;
                    }

                    int next_cost = cost + graph[nr][nc];

                    if (visited[nr][nc] <= next_cost) {
                        continue;
                    }

                    visited[nr][nc] = next_cost;

                    que.add(new int[] {nr, nc, next_cost});
                }
            }

            System.out.println("Problem " + p +": " + visited[n-1][n-1]);

        }
    }

}
